packages/0x.js/src/order_watcher/remaining_fillable_calculator.ts

```diff
@@ -0,0 +1,71 @@
+        if (this.hasSufficientFundsForFeeAndTransferAmount()) {
+            return this._remainingMakerTokenAmount;
+        }
+        if (this._signedOrder.makerFee.isZero()) {
+            return BigNumber.min(this._remainingMakerTokenAmount, this._transferrableMakerTokenAmount);
+        } else {
+            return this.calculatePartiallyFillableMakerTokenAmount();
+        }
+    }
+    public computeRemainingTakerFillable(): BigNumber {
+        return this.computeRemainingMakerFillable().times(this._signedOrder.takerTokenAmount)
+                                                   .dividedToIntegerBy(this._signedOrder.makerTokenAmount);
+    }
+    private hasSufficientFundsForFeeAndTransferAmount(): boolean {
+        if (this._isMakerTokenZRX) {
+            const totalZRXTransferAmountRequired = this._remainingMakerTokenAmount.plus(this._remainingMakerFeeAmount);
+            return this._transferrableMakerTokenAmount.gte(totalZRXTransferAmountRequired);
+        } else {
+            const hasSufficientFundsForTransferAmount = this._transferrableMakerTokenAmount.gte(
+                                                            this._remainingMakerTokenAmount);
+            const hasSufficientFundsForFeeAmount = this._transferrableMakerFeeTokenAmount.gte(
+                                                       this._remainingMakerFeeAmount);
+            return (hasSufficientFundsForTransferAmount && hasSufficientFundsForFeeAmount);
+        }
+    }
+
+    private calculatePartiallyFillableMakerTokenAmount(): BigNumber {
+        const orderToFeeRatio = this._signedOrder.makerTokenAmount.dividedToIntegerBy(this._signedOrder.makerFee);
+        // Maximum number of times the Maker can fill the order, given the fees
+        const fillableTimesInFeeTokenUnits = BigNumber.min(this._transferrableMakerFeeTokenAmount,
+                                                           this._remainingMakerFeeAmount);
+        // Maximum number of times the Maker can fill the order, given the Maker Token Balance
+        let fillableTimesInMakerTokenUnits = this._transferrableMakerTokenAmount.dividedToIntegerBy(orderToFeeRatio);
+        if (this._isMakerTokenZRX) {
+            const totalZRXTokenPooled = this._transferrableMakerTokenAmount;
+            // The purchasing power here is less as the tokens are taken from the same Pool
+            // For every one number of fills, we have to take an extra ZRX out of the pool
+            fillableTimesInMakerTokenUnits = totalZRXTokenPooled.dividedToIntegerBy(
+                                                                     orderToFeeRatio.plus(new BigNumber(1)));
+
+        }
+        const partiallyFillableMakerTokenAmount = fillableTimesInMakerTokenUnits.times(orderToFeeRatio);
+        const partiallyFillableFeeTokenAmount = fillableTimesInFeeTokenUnits.times(orderToFeeRatio);
+        return BigNumber.min(partiallyFillableMakerTokenAmount, partiallyFillableFeeTokenAmount);
```

Comment:
We have a rule to never return an expression. Assign it to a variable and return a variable.